applications/main/u2f/u2f.c

```diff
@@ -327,22 +334,50 @@ static uint16_t u2f_authenticate(U2fData* U2F, uint8_t* buf) {
     return (sizeof(U2fAuthResp) + signature_len + 2);
 }
 
-uint16_t u2f_msg_parse(U2fData* U2F, uint8_t* buf, uint16_t len) {
-    furi_assert(U2F);
-    if(!U2F->ready) return 0;
-    if((buf[0] != 0x00) && (len < 5)) return 0;
-    if(buf[1] == U2F_CMD_REGISTER) { // Register request
-        return u2f_register(U2F, buf);
+uint16_t u2f_applet_selection(const uint8_t* in_buf, uint8_t* out_buf) {
+    U2fApduCommand* cmd = (U2fApduCommand*)in_buf;
+    uint8_t* data = apdu_command_data(cmd);
+
+    FURI_LOG_D(
+        TAG,
+        "len=%d %02x%02x%02x%02x%02x%02x%02x%02x",
+        cmd->len[0],
+        data[0],
+        data[1],
+        data[2],
+        data[3],
+        data[4],
+        data[5],
+        data[6],
+        data[7]);
+
+    if(cmd->len[0] != 8 || memcmp(rid_ac_ax, data, 8) != 0) {
+        memcpy(&out_buf[0], state_app_not_found, 2);
+        return 2;
+    }
 
-    } else if(buf[1] == U2F_CMD_AUTHENTICATE) { // Authenticate request
-        return u2f_authenticate(U2F, buf);
+    memcpy(&out_buf[0], ver_str, 6);
+    memcpy(&out_buf[6], state_no_error, 2);
+    return 8;
+}
 
-    } else if(buf[1] == U2F_CMD_VERSION) { // Get U2F version string
-        memcpy(&buf[0], ver_str, 6);
-        memcpy(&buf[6], state_no_error, 2);
+uint16_t u2f_msg_parse(U2fData* U2F, const uint8_t* in_buf, uint16_t in_len, uint8_t* out_buf) {
+    furi_assert(U2F);
+    if(!U2F->ready) return 0;
+    if((in_buf[0] != 0x00) && (in_len < 5)) return 0;
+    FURI_LOG_D(TAG, "ins=0x%02x", in_buf[1]);
+    if(in_buf[1] == U2F_CMD_REGISTER) { // Register request
+        return u2f_register(U2F, in_buf, out_buf);
+    } else if(in_buf[1] == U2F_CMD_AUTHENTICATE) { // Authenticate request
+        return u2f_authenticate(U2F, in_buf, out_buf);
+    } else if(in_buf[1] == U2F_CMD_VERSION) { // Get U2F version string
+        memcpy(&out_buf[0], ver_str, 6);
+        memcpy(&out_buf[6], state_no_error, 2);
         return 8;
+    } else if(in_buf[1] == U2F_CMD_APPLET_SELECTION) {
+        return u2f_applet_selection(in_buf, out_buf);
```

Comment:
This doesn't appear to update any global state, so this makes it possible to communicate with the U2F applet **without** `SELECT`, which is [explicitly disallowed by the CTAP 2.x spec][0] (emphasis mine):

> * NFC, when the applet is selected [as specified](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#nfc-applet-selection).
>   * **Authenticator SHALL NOT allow FIDO applet to be implicitly selected or enabled.**
>     * Recommended: Authenticator SHALL NOT have default applet selected on power cycle. All CTAP commands SHALL be preceded by an explicit applet selection command as described in [Applet selection](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#nfc-applet-selection) section.
>      * Alternative: If authenticator has a FIDO applet selected for some reason at power cycle, it SHALL be in disabled mode and SHALL ONLY be enabled once it receives explicit applet selection command as described in [Applet selection](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#nfc-applet-selection) section.

[The U2F 1.2 NFC spec][2] also states that applet selection is required:

> A FIDO client SHALL always send an applet selection command to begin interaction with a FIDO authenticator via NFC.

[Some operating systems filter SELECT commands sent over smartcard interfaces][3] in order to limit direct access to the CTAP/U2F applet; so any device which allows implicit selection would bypass this control.

The [deselect command][1] is also not implemented by this PR, but that's not part of the U2F 1.2 spec, so it may be worth waiting for proper CTAP2 support for that one.

[0]: https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#sctn-secure-interaction
[1]: https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#nfc-applet-deselect
[2]: https://fidoalliance.org/specs/fido-u2f-v1.2-ps-20170411/fido-u2f-nfc-protocol-v1.2-ps-20170411.html#applet-selection
[3]: https://learn.microsoft.com/en-us/windows/security/identity-protection/hello-for-business/webauthn-apis